File: mrbgems/mruby-digest/src/digest.c
/*
** digest.c - Digest module
**
** See Copyright Notice in mruby.h
*/

#include <mruby.h>
#include <mruby/array.h>
#include <mruby/class.h>
#include <mruby/data.h>
#include <mruby/digest.h>
#include <mruby/internal.h>
#include <mruby/string.h>
#include <mruby/variable.h>

#include <string.h>

/* Where a class keeps the algorithm its instances compute. It is read by
   walking up from an instance's class, so a subclass written in Ruby
   computes what its superclass computes without being given anything. The
   name is one no Ruby program can write, since the walk has to tell a class
   that was given an algorithm from one that merely carries an instance
   variable of its own. */
#define DIGEST_METADATA_IV MRB_SYM(mruby_Digest_metadata)

typedef struct digest_object {
  const mrb_digest_metadata *md;
  /* The algorithm's state, in an allocation of its own. Carrying it here
     rather than behind this pointer would put a flexible array member in a
     struct that also has to compile as C++, and would leave the alignment of
     the state to be arranged by hand. */
  void *ctx;
} digest_object;

static void
digest_free(mrb_state *mrb, void *ptr)
{
  digest_object *d = (digest_object*)ptr;

  if (d) {
    mrb_free(mrb, d->ctx);
    mrb_free(mrb, d);
  }
}

static const mrb_data_type digest_type = { "Digest", digest_free };

/* The algorithm the given class computes, or NULL where no class in the
   chain was given one: that is Digest::Base itself, and any subclass of it
   that names no algorithm. */
static const mrb_digest_metadata*
digest_metadata(mrb_state *mrb, struct RClass *c)
{
  for (; c; c = c->super) {
    /* An included module reaches the chain as an iclass, which shares its
       instance variables with the module it stands for; skipping it keeps
       the walk to the classes that were actually given an algorithm. */
    if (c->tt == MRB_TT_ICLASS) continue;

    mrb_value v = mrb_iv_get(mrb, mrb_obj_value(c), DIGEST_METADATA_IV);
    if (mrb_cptr_p(v)) return (const mrb_digest_metadata*)mrb_cptr(v);
  }
  return NULL;
}

static digest_object*
digest_ptr(mrb_state *mrb, mrb_value self)
{
  digest_object *d = (digest_object*)mrb_data_get_ptr(mrb, self, &digest_type);

  /* A digest object with no state is one whose #initialize was never
     reached, which allocate alone can arrive at. */
  if (d == NULL || d->ctx == NULL) {
    mrb_raise(mrb, E_RUNTIME_ERROR, "uninitialized digest object");
  }
  return d;
}

/* The state of a digest object is what its methods write, so a frozen one
   refuses every method that would move it along. What #digest does to a
   frozen object it does to an unfrozen copy of it, so that still works. */
static digest_object*
digest_ptr_for_write(mrb_state *mrb, mrb_value self)
{
  mrb_check_frozen_value(mrb, self);
  return digest_ptr(mrb, self);
}

/* Gives self a fresh context for the algorithm md, dropping whatever it
   held. The half-built object is attached before the second allocation, so
   a failure there leaves it to be freed rather than lost. */
static digest_object*
digest_alloc(mrb_state *mrb, mrb_value self, const mrb_digest_metadata *md)
{
  mrb_check_frozen_value(mrb, self);

  /* #initialize and #initialize_copy are ordinary methods and can be
     reached again on an object that already holds a context. */
  if (DATA_TYPE(self) == &digest_type) {
    digest_free(mrb, DATA_PTR(self));
    mrb_data_init(self, NULL, &digest_type);
  }

  digest_object *d = (digest_object*)mrb_malloc(mrb, sizeof(digest_object));

  d->md = md;
  d->ctx = NULL;
  mrb_data_init(self, d, &digest_type);
  d->ctx = mrb_malloc(mrb, md->context_size);
  return d;
}

/*
 * call-seq:
 *   Digest::SHA256.new -> digest_obj
 *
 * Returns a digest object with nothing fed into it yet.
 */
static mrb_value
digest_base_init(mrb_state *mrb, mrb_value self)
{
  const mrb_digest_metadata *md = digest_metadata(mrb, mrb_obj_class(mrb, self));

  if (md == NULL) {
    mrb_raisef(mrb, E_NOTIMP_ERROR, "%C names no digest algorithm", mrb_obj_class(mrb, self));
  }

  digest_object *d = digest_alloc(mrb, self, md);
  md->init(d->ctx);
  return self;
}

/*
 * Copies what the source has been fed so far, so that #dup and #clone hand
 * back an object that continues from the same point rather than one sharing
 * the original's state.
 */
static mrb_value
digest_base_init_copy(mrb_state *mrb, mrb_value self)
{
  mrb_value src = mrb_get_arg1(mrb);

  if (mrb_obj_equal(mrb, self, src)) return self;

  digest_object *s = digest_ptr(mrb, src);
  digest_object *d = digest_alloc(mrb, self, s->md);
  memcpy(d->ctx, s->ctx, s->md->context_size);
  return self;
}

/*
 * call-seq:
 *   digest_obj.update(string) -> digest_obj
 *   digest_obj << string      -> digest_obj
 *
 * Feeds the string's bytes to the digest and returns the receiver, so that
 * the calls chain.
 *
 *   Digest::SHA256.new << "ab" << "c"
 */
static mrb_value
digest_base_update(mrb_state *mrb, mrb_value self)
{
  mrb_value str;
  mrb_get_args(mrb, "S", &str);

  digest_object *d = digest_ptr_for_write(mrb, self);
  d->md->update(d->ctx, RSTRING_PTR(str), (size_t)RSTRING_LEN(str));
  return self;
}

/*
 * call-seq:
 *   digest_obj.reset -> digest_obj
 *
 * Puts the digest back to where a fresh one stands.
 */
static mrb_value
digest_base_reset(mrb_state *mrb, mrb_value self)
{
  digest_object *d = digest_ptr_for_write(mrb, self);

  d->md->init(d->ctx);
  return self;
}

/*
 * Finishes the digest and returns its bytes. The context is left finished,
 * which is why the callers in Digest::Instance either work on a copy or
 * reset afterwards.
 */
static mrb_value
digest_base_finish(mrb_state *mrb, mrb_value self)
{
  digest_object *d = digest_ptr_for_write(mrb, self);
  mrb_value str = mrb_str_new(mrb, NULL, d->md->digest_length);

  /* A digest is bytes rather than text, so the string it comes back in says
     so. Without this the bytes would be read as the build's default
     encoding, where all but one digest in eight is broken UTF-8. */
  RSTR_ENCODING_SET(mrb_str_ptr(str), MRB_STR_ENCODING_BINARY);
  d->md->finish(d->ctx, (unsigned char*)RSTRING_PTR(str));
  return str;
}

/*
 * call-seq:
 *   digest_obj.digest_length -> integer
 *
 * The length in bytes of the digests this object produces.
 */
static mrb_value
digest_base_digest_length(mrb_state *mrb, mrb_value self)
{
  digest_object *d = digest_ptr(mrb, self);

  return mrb_int_value(mrb, (mrb_int)d->md->digest_length);
}

/*
 * call-seq:
 *   digest_obj.block_length -> integer
 *
 * The length in bytes of the input block the algorithm works on.
 */
static mrb_value
digest_base_block_length(mrb_state *mrb, mrb_value self)
{
  digest_object *d = digest_ptr(mrb, self);

  return mrb_int_value(mrb, (mrb_int)d->md->block_length);
}

static const char digest_hexmap[] = "0123456789abcdef";

static mrb_value
digest_hexencode(mrb_state *mrb, mrb_value src)
{
  mrb_int len = RSTRING_LEN(src);

  if (len > MRB_INT_MAX / 2) {
    mrb_raise(mrb, E_ARGUMENT_ERROR, "string too long to hex-encode");
  }

  mrb_value dst = mrb_str_new(mrb, NULL, len * 2);
  const unsigned char *p = (const unsigned char*)RSTRING_PTR(src);
  char *q = RSTRING_PTR(dst);

  for (mrb_int i = 0; i < len; i++) {
    *q++ = digest_hexmap[p[i] >> 4];
    *q++ = digest_hexmap[p[i] & 0x0f];
  }
  /* Every byte written above is ASCII, so the walk that would find that out
     is answered here instead. */
  RSTR_CODERANGE_SET(mrb_str_ptr(dst), MRB_STR_CODERANGE_7BIT);
  return dst;
}

/*
 * call-seq:
 *   Digest.hexencode(string) -> string
 *
 * Returns the string's bytes spelled as lowercase hexadecimal, which is
 * what #hexdigest hands back a digest in.
 *
 *   Digest.hexencode("\x01\xff")  #=> "01ff"
 */
static mrb_value
digest_s_hexencode(mrb_state *mrb, mrb_value self)
{
  mrb_value str;
  mrb_get_args(mrb, "S", &str);

  return digest_hexencode(mrb, str);
}

static const mrb_mt_entry digest_base_rom_entries[] = {
  MRB_MT_ENTRY(digest_base_init,          MRB_SYM(initialize),      MRB_ARGS_NONE()),
  MRB_MT_ENTRY(digest_base_init_copy,     MRB_SYM(initialize_copy), MRB_ARGS_REQ(1)),
  MRB_MT_ENTRY(digest_base_update,        MRB_SYM(update),          MRB_ARGS_REQ(1)),
  MRB_MT_ENTRY(digest_base_update,        MRB_OPSYM(lshift),        MRB_ARGS_REQ(1)),
  MRB_MT_ENTRY(digest_base_reset,         MRB_SYM(reset),           MRB_ARGS_NONE()),
  /* Private for the reason CRuby's is: what a digest object is asked for is
     a digest, and #finish leaves the object somewhere only its own methods
     know how to carry on from. */
  MRB_MT_ENTRY(digest_base_finish,        MRB_SYM(finish),          MRB_ARGS_NONE() | MRB_MT_PRIVATE),
  MRB_MT_ENTRY(digest_base_digest_length, MRB_SYM(digest_length),   MRB_ARGS_NONE()),
  MRB_MT_ENTRY(digest_base_block_length,  MRB_SYM(block_length),    MRB_ARGS_NONE()),
};

MRB_API struct RClass*
mrb_digest_define_algorithm(mrb_state *mrb, const char *name, const mrb_digest_metadata *md)
{
  struct RClass *digest = mrb_module_get_id(mrb, MRB_SYM(Digest));
  struct RClass *base = mrb_class_get_under_id(mrb, digest, MRB_SYM(Base));
  struct RClass *c = mrb_define_class_under(mrb, digest, name, base);

  mrb_iv_set(mrb, mrb_obj_value(c), DIGEST_METADATA_IV, mrb_cptr_value(mrb, (void*)md));
  return c;
}

/* Defined in the file of the algorithm it registers. */
void mrb_digest_sha256_init(mrb_state *mrb);

void
mrb_mruby_digest_gem_init(mrb_state *mrb)
{
  struct RClass *digest = mrb_define_module_id(mrb, MRB_SYM(Digest));
  mrb_define_module_function_id(mrb, digest, MRB_SYM(hexencode), digest_s_hexencode, MRB_ARGS_REQ(1));

  /* The three of them stand where CRuby's do: Instance carries what every
     digest object answers, Class is what a digest class is, and Base is the
     one whose instances hold a context. mrblib/digest.rb fills the first two
     in; it runs after this, which is why the classes themselves are made
     here rather than there. */
  struct RClass *instance = mrb_define_module_under_id(mrb, digest, MRB_SYM(Instance));
  struct RClass *klass = mrb_define_class_under_id(mrb, digest, MRB_SYM(Class), mrb->object_class);
  mrb_include_module(mrb, klass, instance);

  struct RClass *base = mrb_define_class_under_id(mrb, digest, MRB_SYM(Base), klass);
  MRB_SET_INSTANCE_TT(base, MRB_TT_CDATA);
  MRB_MT_INIT_ROM(mrb, base, digest_base_rom_entries);

  mrb_digest_sha256_init(mrb);
}

void
mrb_mruby_digest_gem_final(mrb_state *mrb)
{
}

File: mrbgems/mruby-digest/test/digest.rb
##
# Digest Test

# FIPS 180-4 and the NIST example set. The empty message and the 56-byte one
# are the two cases the padding is written for: nothing to pad but the block
# itself, and a length that no longer fits in the block it ends.
SHA256_VECTORS = {
  "" => "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
  "abc" => "ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad",
  "abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq" =>
    "248d6a61d20638b8e5c026930c3e6039a33ce45964ff2167f6ecedd419db06c1",
  "abcdefghbcdefghicdefghijdefghijkefghijklfghijklmghijklmnhijklmnoijklmnopjklmnopqklmnopqrlmnopqrsmnopqrstnopqrstu" =>
    "cf5b16a778af8380036ce59e7b0492370b249b11e8f07a51afac45037afee9d1",
}

assert('Digest::SHA256 test vectors') do
  SHA256_VECTORS.each do |message, want|
    assert_equal want, Digest::SHA256.hexdigest(message)
    assert_equal want, Digest::SHA256.new.update(message).hexdigest
  end
end

assert('Digest::SHA256 a message of a million bytes') do
  # The one vector long enough to run the block loop rather than the buffer,
  # fed in pieces that do not line up with a block so that both are used.
  d = Digest::SHA256.new
  1000.times { d << "a" * 1000 }
  assert_equal "cdc76e5c9914fb9281a1c7e284d73e67f1809a48a497200e046d39ccc7112cd0", d.hexdigest
end

assert('Digest::SHA256 a message fed one byte at a time') do
  # Whatever the buffering does with a message, it has to arrive at what the
  # message digests to whole. The lengths are the ones the block boundaries
  # fall on, either side of them, and where the length no longer fits.
  [0, 1, 54, 55, 56, 57, 63, 64, 65, 119, 120, 128, 129].each do |len|
    message = "0123456789" * 13
    message = message[0, len]
    d = Digest::SHA256.new
    len.times { |i| d << message[i, 1] }
    assert_equal Digest::SHA256.hexdigest(message), d.hexdigest, "length #{len}"
  end
end

assert('Digest::SHA256#digest returns the bytes') do
  digest = Digest::SHA256.digest("abc")
  assert_equal 32, digest.bytesize
  assert_equal Digest::SHA256.hexdigest("abc"), Digest.hexencode(digest)
end

assert('Digest::SHA256#digest leaves the object where it found it') do
  d = Digest::SHA256.new
  d << "ab"
  # Given a string, the digest is that string's and the object is emptied;
  # given nothing, the object carries on from where it stood.
  assert_equal Digest::SHA256.hexdigest("xyz"), Digest.hexencode(d.digest("xyz"))
  assert_equal Digest::SHA256.hexdigest(""), d.hexdigest
  d << "ab"
  assert_equal Digest::SHA256.hexdigest("ab"), d.hexdigest
  d << "c"
  assert_equal Digest::SHA256.hexdigest("abc"), d.hexdigest
end

assert('Digest::SHA256#digest! empties the object') do
  d = Digest::SHA256.new
  d << "abc"
  assert_equal Digest::SHA256.digest("abc"), d.digest!
  assert_equal Digest::SHA256.hexdigest(""), d.hexdigest
end

assert('Digest::SHA256#hexdigest!') do
  d = Digest::SHA256.new
  d << "abc"
  assert_equal Digest::SHA256.hexdigest("abc"), d.hexdigest!
  assert_equal Digest::SHA256.hexdigest(""), d.hexdigest
end

assert('Digest::SHA256#reset') do
  d = Digest::SHA256.new
  d << "abc"
  assert_same d, d.reset
  assert_equal Digest::SHA256.hexdigest(""), d.hexdigest
end

assert('Digest::SHA256#dup carries the state and shares none of it') do
  d = Digest::SHA256.new
  d << "ab"
  copy = d.dup
  copy << "c"
  d << "z"
  assert_equal Digest::SHA256.hexdigest("abc"), copy.hexdigest
  assert_equal Digest::SHA256.hexdigest("abz"), d.hexdigest
end

assert('Digest::SHA256#new returns an empty digest of the same class') do
  d = Digest::SHA256.new
  d << "abc"
  fresh = d.new
  assert_equal Digest::SHA256, fresh.class
  assert_equal Digest::SHA256.hexdigest(""), fresh.hexdigest
  assert_equal Digest::SHA256.hexdigest("abc"), d.hexdigest
end

assert('Digest::SHA256 lengths') do
  d = Digest::SHA256.new
  assert_equal 32, d.digest_length
  assert_equal 32, d.length
  assert_equal 32, d.size
  assert_equal 64, d.block_length
end

assert('Digest::SHA256#== and #to_s') do
  d = Digest::SHA256.new.update("abc")
  other = Digest::SHA256.new.update("abc")
  assert_true d == other
  assert_true d == Digest::SHA256.hexdigest("abc")
  assert_equal Digest::SHA256.hexdigest("abc"), d.to_s
  other << "d"
  assert_false d == other
  assert_false d == "not a digest"
  assert_false d == 1
end

assert('Digest::SHA256#inspect') do
  hex = Digest::SHA256.hexdigest("abc")
  d = Digest::SHA256.new.update("abc")
  assert_equal "#<Digest::SHA256: #{hex}>", d.inspect
end

assert('Digest::SHA256#update takes a String and nothing else') do
  d = Digest::SHA256.new
  assert_raise(TypeError) { d << 1 }
  assert_raise(TypeError) { d.update(nil) }
end

assert('Digest::SHA256#finish is private') do
  d = Digest::SHA256.new
  assert_raise(NoMethodError) { d.finish }
end

assert('a frozen digest refuses what would move it along') do
  d = Digest::SHA256.new
  d << "abc"
  d.freeze
  assert_raise(FrozenError) { d << "d" }
  assert_raise(FrozenError) { d.update("d") }
  assert_raise(FrozenError) { d.reset }
  assert_raise(FrozenError) { d.digest! }
  # Reading it is still reading: what these two finish is a copy.
  assert_equal Digest::SHA256.hexdigest("abc"), d.hexdigest
  assert_equal Digest::SHA256.digest("abc"), d.digest
  copy = d.clone
  assert_true copy.frozen?
  assert_equal Digest::SHA256.hexdigest("abc"), copy.hexdigest
  copy = d.dup
  assert_false copy.frozen?
  copy << "d"
  assert_equal Digest::SHA256.hexdigest("abcd"), copy.hexdigest
end

assert('Digest::Base names no algorithm') do
  assert_raise(NotImplementedError) { Digest::Base.new }
end

assert('a digest class subclassed in Ruby computes what it inherits') do
  c = Class.new(Digest::SHA256)
  assert_equal Digest::SHA256.hexdigest("abc"), c.hexdigest("abc")
  assert_equal 32, c.new.digest_length
end

assert('Digest.hexencode') do
  assert_equal "", Digest.hexencode("")
  assert_equal "0001ff80", Digest.hexencode("\x00\x01\xff\x80")
  assert_equal "616263", Digest.hexencode("abc")
  assert_raise(TypeError) { Digest.hexencode(1) }
end

assert('a digest is bytes rather than text') do
  # A digest holds whatever the algorithm produced, which is why it says it
  # is bytes rather than letting them be read as the build's default
  # encoding. The digest of "1" is what shows the difference: two of its
  # bytes read as one UTF-8 character, so a build that indexes by character
  # would count 31 of them in the 32 bytes.
  d = Digest::SHA256.digest("1")
  assert_equal 32, d.bytesize
  assert_equal 32, d.length

  # Where the build carries encodings as well, the digest names the one it
  # is in. mruby-encoding is not among this gem's test dependencies, since
  # it is what turns a build's strings from bytes into UTF-8 and pulling it
  # in would take that choice away from the build; the assertions below run
  # wherever a state carries both gems.
  if d.respond_to?(:encoding)
    assert_equal Encoding::BINARY, d.encoding
    assert_true d.valid_encoding?
    assert_equal Encoding::UTF_8, Digest::SHA256.hexdigest("1").encoding
  end
end
